### PR TITLE
feat(linear-logp): add batch-invariant fused SM90 operator

### PR DESCRIPTION
**Pull request by ThreeMonth03**
The existing SM90 fused `linear_logp` avoids materializing `[N, V]` logits, but its throughput scheduler chooses the vocabulary split from `N`, `V`, and the GPU SM count. The same token row can therefore take a different FP32 reduction path when its batch layout changes.

## Why this happens

`A`–`D` below are consecutive chunks of the same vocabulary. The token row, weight, target, and vocabulary are unchanged; only `N` changes.

```mermaid
flowchart TB
    X["Same token row, weight, target, and vocabulary"]
    X --> N1["Run inside N = 1"]
    X --> N2["Run inside N = 1024"]

    N1 --> S1["Throughput scheduler may choose<br/>[A] [B] [C] [D]"]
    N2 --> S2["Throughput scheduler may choose<br/>[A B] [C D]"]

    S1 --> R1["Merge four partial<br/>online-softmax states"]
    S2 --> R2["Merge two grouped partial<br/>online-softmax states"]

    R1 --> F["Different FP32 rounding<br/>may change the output bits"]
    R2 --> F
```

Both paths are equivalent in real arithmetic. They need not be bitwise identical because FP32 renormalization and addition are not associative.

## What this PR changes

```mermaid
flowchart LR
    subgraph Added["Added by this PR"]
        API["batch_invariant_linear_logp<br/>new entry point"] --> POLICY["Choose split-V from V only<br/>new policy"]
    end

    POLICY --> FIXED["Same contiguous vocabulary splits<br/>for every N"]

    subgraph Reused["Reused from existing linear_logp"]
        KERNEL["TMA + MMA projection kernel"] --> MERGE["Online-softmax partials<br/>and ascending combine"]
    end

    FIXED --> KERNEL
    MERGE --> OUT["Same row + same V<br/>same arithmetic order and bits"]

    N["N"] --> ROWS["Changes row CTA count only<br/>does not select split-V"]
    ROWS -.-> KERNEL
```

The existing throughput-oriented `linear_logp` remains available. This PR adds an opt-in invariant entry point and a `V`-only split policy; it reuses the existing projection, online-softmax, and combine kernels.

| Scope | Included |
| --- | --- |
| Hardware | Single-card SM90 Hopper |
| Dtype | BF16 hidden states and weights |
| Execution | Forward only |
| Excluded | Tensor parallelism, backward, fallback backends |

Tests cover FP32 reference accuracy and exact equality across batch size, row position, neighboring rows, chunking, layouts, vocabulary boundaries, and CUDA streams. Local CUDA 13.1 SM90a compilation and non-H100 checks pass; H100 GPU CI is still required before this leaves draft.

Related to #122. Potentially overlaps with the broader strict mode in #336.

## Summary by CodeRabbit

- **New Features**
  - Added a batch-invariant fused linear log-probability operator for Hopper (SM90) GPUs.
  - Supports deterministic results across batch layouts without materializing logits.
  - Added device-aware dispatch, input validation, and fail-closed behavior when unsupported.

- **Documentation**
  - Added operator documentation, navigation links, API details, and benchmarking guidance.
  - Clarified that the standard fused operator prioritizes throughput over bitwise batch invariance.

- **Tests & Benchmarks**
  - Added extensive correctness, validation, dispatch, stream, alignment, and regression coverage.
  - Added a benchmark comparing fused, batch-invariant, and materialized computation paths.

### CODE REVIEW

File: csrc/cuda/fused_linear_logp_sm90.cu [hunk 1]
```diff
@@ -54,13 +57,76 @@ constexpr int GRAD_W_WMMA_K = 16;
...
 static_assert(BK % 32 == 0, "BK must be a multiple of 32 (ldmatrix.x4 spans 32 cols)");
 
 inline void init_tensor_map_noswizzle(CUtensorMap *tmap, const nv_bfloat16 *gmem,
                                       uint64_t gmem_height, uint64_t gmem_width,
                                       uint32_t box_height, uint32_t box_width);
 
+int select_batch_invariant_vocab_splits(int total_vtiles) {
+    // Keep at most MAX_BATCH_INVARIANT_VOCAB_SPLITS non-empty, contiguous
+    // ranges. Both the partition and the combine order depend on V only.
+    const int vtiles_per_split =
```
Comment (ThreeMonth03) [inline csrc/cuda/fused_linear_logp_sm90.cu:80]: `V -> split count -> contiguous ranges -> ascending merge`. No `N`, row-block, or SM-count input is allowed here; adding one breaks bitwise batch invariance.

File: csrc/cuda/fused_linear_logp_sm90.cu [hunk 2]
```diff
@@ -1814,48 +1948,64 @@ std::vector<torch::Tensor> fused_linear_logp_sm90_forward_impl(
...
+                                               VocabSplitPolicy::kThroughput);
+}
+
+// Single-card batch-invariant forward. It deliberately shares the validated
+// TMA+MMA device kernel with the throughput path, but pins split-V independently
+// of N so each row sees the same D and V reduction topology across batch layouts.
+std::vector<torch::Tensor> batch_invariant_linear_logp_sm90_forward(
+    torch::Tensor hidden, torch::Tensor weight, torch::Tensor target,
+    torch::optional<torch::Tensor> bias) {
+    return fused_linear_logp_sm90_forward_impl(hidden, weight, target, bias, 0, false,
```
Comment (ThreeMonth03) [inline csrc/cuda/fused_linear_logp_sm90.cu:2013]: `linear_logp -> kThroughput`; `batch_invariant_linear_logp -> kBatchInvariant`. Both reuse the validated kernel, so existing callers keep their occupancy-tuned behavior.

File: rl_engine/kernels/ops/cuda/loss/batch_invariant_linear_logp.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            raise ValueError("bias must be a CUDA tensor on the same device as hidden")
+        if bias.dtype not in _BIAS_DTYPES:
+            raise TypeError(f"bias must have dtype fp16, bf16, or fp32, got {bias.dtype}")
+
+    if torch.is_grad_enabled() and (
+        hidden.requires_grad
+        or lm_head_weight.requires_grad
+        or (bias is not None and bias.requires_grad)
+    ):
+        raise RuntimeError(
```
Comment (ThreeMonth03) [inline rl_engine/kernels/ops/cuda/loss/batch_invariant_linear_logp.py:96]: `requires_grad -> error`; `no_grad -> SM90 forward`. Backward is outside the first contract, so this path fails closed.

File: rl_engine/kernels/registry.py [hunk 4]
```diff
@@ -280,6 +289,7 @@ def __init__(self):
                 "grpo_loss": [OpBackend.PYTORCH_GRPO_LOSS],
                 "rope": [OpBackend.PYTORCH_NATIVE_ROPE],
                 "linear_logp": [OpBackend.PYTORCH_LINEAR_LOGP],
+                "batch_invariant_linear_logp": [],
```
Comment (ThreeMonth03) [inline rl_engine/kernels/registry.py:292]: `unsupported platform -> [] -> RuntimeError`. If this key were absent, the registry would choose its generic PyTorch default.

File: scripts/ci_smoke.py [hunk 5]
```diff
@@ -70,7 +71,49 @@ def main() -> int:
         )
         return 1
 
-    print(f"[smoke] OK: rl_engine._C built and fused_logp ran on sm_{cc[0]}{cc[1]}.")
+    if cc[0] == 9:
```
Comment (ThreeMonth03) [inline scripts/ci_smoke.py:74]: `H100 build -> symbol check -> launch -> synchronize`. Generic `fused_logp` cannot detect an omitted conditionally compiled SM90 feature.

File: tests/test_batch_invariant_linear_logp.py [hunk 6]
```diff
@@ -0,0 +1,768 @@
...
+                op(
+                    hidden[start : start + chunk_size],
+                    weight,
+                    target_ids[start : start + chunk_size],
+                    bias,
+                )
+                for start in range(0, hidden.size(0), chunk_size)
+            ]
+        )
+        assert torch.equal(chunked, full), f"batch chunk size {chunk_size} changed output bits"
```
Comment (ThreeMonth03) [inline tests/test_batch_invariant_linear_logp.py:442]: `full-batch bits == concat(chunk bits)`. `torch.equal` enforces the contract; `allclose` would hide the regression.

**Review by ThreeMonth03 — COMMENTED** [review]
@KJLdefeated @inaniloquentee please review this pull request.

**Comment by ThreeMonth03** [comment]
Closing because #336 has been merged and now provides the strict bitwise batch-invariant SM90 linear-logp contract, including single-rank/TP execution, temperature, padded-vocab masking, returned logits, and backward support. This supersedes the original scope of this PR. Any remaining alignment or input-validation hardening will be submitted separately as focused fixes.
